File: Grind75/45-MergeIntervals/cpp/MergeIntervals.cpp
#include "MergeIntervals.h"
#include <algorithm>
#include <list>
#include <vector>
#include <prettyPrint.h>

using namespace std;

inline bool compare(const vector<int> &left, const vector<int> &right) {
  return left[0] < right[0];
}

vector<vector<int>> Solution::merge(vector<vector<int>>& intervals) {
  if (intervals.size() < 2)
    return intervals;
  sort(intervals.begin(), intervals.end(), compare);
  vector<vector<int>> result;
  result.push_back(intervals.front());
  for (size_t i = 1; i < intervals.size(); i++) {
    vector<int> &iv = intervals[i];
    vector<int> &prev = result.back();
    if (iv[0] <= prev[1])
      prev[1] = max(iv[1], prev[1]);
    else
      result.push_back(iv);
  }
  return result;
}

vector<vector<int>> Solution::merge2(vector<vector<int>>& intervals) {
  if (intervals.size() < 2)
    return intervals;
  sort(intervals.begin(), intervals.end(), compare);
  list<vector<int>> iList;
  iList.insert(iList.begin(), intervals.begin(), intervals.end());
  list<vector<int>>::iterator prev = iList.begin();
  list<vector<int>>::iterator it = ++iList.begin();
  while (it != iList.end()) {
    if ((*it)[0] <= (*prev)[1]) {
      (*prev)[1] = max((*it)[1], (*prev)[1]);
      it = iList.erase(it);
    } else {
      it++;
      prev++;
    }
  }
  vector<vector<int>> result;
  result.insert(result.begin(), iList.begin(), iList.end());
  return result;
}

File: Grind75/45-MergeIntervals/python/MergeIntervals.py
#!/bin/python3

from __future__ import annotations
import sys
sys.path.append('../../Utility/python')
from typing import List, Tuple
from PrettyPrint import *
from Colors import *

class Solution(object):
  def merge(self, intervals: List[List[int]]) -> List[List[int]]:
    if len(intervals) < 2:
      return intervals
    intervals.sort(key=lambda x: x[0])
    result: List[List[int]] = [intervals[0]]
    for i in range(1, len(intervals)):
      iv = intervals[i]
      prev = result[-1]
      if iv[0] <= prev[1]:
        prev[1] = max(prev[1], iv[1])
      else:
        result.append(iv)
    return result

if __name__ == '__main__':
  soln = Solution()
  testCases: List[Tuple[ List[List[int]], List[List[int]] ]] = [
    ( [], [] ),
    ( [[0,1]], [[0,1]] ),
    ( [[0,1],[2,3]], [[0,1],[2,3]] ),
    ( [[0,1],[1,2]], [[0,2]] ),
    ( [[0,1],[1,2]], [[0,2]] ),
    ( [[0,3],[1,2]], [[0,3]] ),
    ( [[1,2],[0,3]], [[0,3]] ),
  ]

  for interavls, merged in testCases:
    colorPrint(LIGHT_BLUE, interavls)
    colorPrint(WHITE, merged)
    result = soln.merge(interavls)
    setMerged = set([tuple(m) for m in merged])
    setResult = set([tuple(r) for r in result])
    colorPrint(GREEN if setResult == setMerged else RED, result)
    pass

